File: Java/src/main/java/hayaa/rpc/client/ServiceMethdoProxy.java
package hayaa.rpc.client;

import hayaa.common.JsonHelper;
import hayaa.common.StringUtil;
import hayaa.rpc.common.config.RPCConfigHelper;
import hayaa.rpc.common.protocol.MethodMessage;
import hayaa.rpc.common.protocol.ResultMessage;

import java.util.HashMap;
import java.util.Hashtable;
import java.util.UUID;

public class ServiceMethdoProxy {
    public static Object invoke(String interfaceName, String methodName, Hashtable<String, Object> paramater,Class<?> resultType) {
        String strResult = null;
        String msgID = UUID.randomUUID().toString();
        MethodMessage methodMessage = new MethodMessage();
        methodMessage.setInterfaceName(interfaceName);
        methodMessage.setMethod(methodName);
        methodMessage.setParamater(paramater);
        methodMessage.setMsgID(msgID);
        ResultMessage msgResult = null;
        int timeOut = RPCConfigHelper.getRPCConfig().getSessionTimeout();
        int time = 0;
        try {

            Boolean action = ClientHelper.get_instance().enQueue(methodMessage);
            if (!action) {
                return null;
            }
            System.out.println("result wait timeOut:"+timeOut);
            while (time < timeOut) {
                System.out.println("result wait loop");
                msgResult = ClientHelper.get_instance().GetResult(msgID);
                if (msgResult != null) {
                    time = timeOut + 1;
                }
                if (time < timeOut) {
                    Thread.sleep(1);
                }
                //考虑代码操作时间增量
                time = time + 3;
            }
            if (StringUtil.IsNullOrEmpty(msgResult.getErrMsg())) {
                strResult = msgResult.getResult();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        System.out.println("Server result:"+strResult);
        Object result=null;
        if(!StringUtil.IsNullOrEmpty(strResult)) {
            result = JsonHelper.gsonDeserialize(strResult, resultType);
        }
        return result;
    }

}

File: C#/RPC/Hayaa.RPC.Componet/Client/ServiceMethdoProxy.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace Hayaa.RPC.Componet
{
    public class ServiceMethdoProxy
    {
        public static ResultMessage Invoke(String interfaceName, String methodName, List<RpcDataValue> paramater)
        {
            ResultMessage result = null;
            try
            {
                String msgID = Guid.NewGuid().ToString("N");               
                int timeOut = RpcClient.GetConfig().SessionTimeout;
                int time = 0;               
                 ClientHelper.Instance.EnQueue(new MethodMessage()
                {
                    InterfaceName = interfaceName,
                    Method = methodName,
                    Paramater = paramater,
                    MsgID = msgID
                });
               // Console.WriteLine("result wait timeOut:" + timeOut);
                while (time<timeOut)
                {
                    Thread.SpinWait(5000);
                    result = ClientHelper.Instance.GetResult(msgID);
                    if (result != null)
                    {
                        Console.WriteLine("time enough");
                    }
                    if (result != null)
                    {
                        time= timeOut+1;
                    }
                    if (time < timeOut)
                        Thread.SpinWait(1);
                    time = time+3;//考虑代码操作时间增量
                }
                if (result == null)
                {
                    Console.WriteLine("remove msgId:" + msgID);
                    ClientHelper.Instance.DelTimeoutMsgID(msgID);
                }               
                Console.WriteLine("Invoke done");
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            return result;
        }

    }
}
